pkg/events/redis/codec.go

```diff
@@ -47,8 +47,11 @@ func decodeEventData(enc string, evt *ttnpb.Event) error {
 	if err != nil {
 		return err
 	}
-	// NOTE: We're merging additional event data into an event that may already contain fields.
-	return proto.UnmarshalMerge(bpb, evt)
+	// NOTE: We override the contents of `evt` with the contents `enc` explicitly.
+	// proto.UnmarshalMerge has append semantics for slices, and as such identifiers
+	// and event visibility will be duplicated, as `evt` most likely contains them
+	// from the sparse event representation.
+	return proto.Unmarshal(bpb, evt)
```

Comment:
Doesn't this unset some of the fields that are only in the metadata, and not in the sparse event? As far as I can tell, the sparse event doesn't contain `name`, `unique_id`, so aren't we un-setting those fields now?

Reply:
The sparse event is `evt`, not `enc`. When we `proto.Unmarshal`, we inherently do a `evt.Reset()`. As such, we are replacing the sparse representation with the full representation.

The function here (`decodeEventData`) has 3 call sites:
- `subscribeTask` - it unmarshals the contents that come via the Redis channels into an empty proto - there is nothing to merge:
https://github.com/TheThingsNetwork/lorawan-stack/blob/127cf5ad55bb0eacda53741ea472a204acf26054/pkg/events/redis/redis.go#L152-L158
- `LoadEvent` - same behavior, load into empty proto without a real merge happening:
https://github.com/TheThingsNetwork/lorawan-stack/blob/127cf5ad55bb0eacda53741ea472a204acf26054/pkg/events/redis/store.go#L110-L121
- `loadEventData` - this is the only one that actually has some partial event representations, but the binary data is complete, as it comes from the event stored by the unique ID:
https://github.com/TheThingsNetwork/lorawan-stack/blob/127cf5ad55bb0eacda53741ea472a204acf26054/pkg/events/redis/store.go#L79-L101

In all 3 cases, the `bpb` here contains a full binary representation of the event - no fields are going to be lost, as we're always storing the full event:
https://github.com/TheThingsNetwork/lorawan-stack/blob/127cf5ad55bb0eacda53741ea472a204acf26054/pkg/events/redis/store.go#L64-L69
https://github.com/TheThingsNetwork/lorawan-stack/blob/127cf5ad55bb0eacda53741ea472a204acf26054/pkg/events/redis/codec.go#L28-L38
https://github.com/TheThingsNetwork/lorawan-stack/blob/127cf5ad55bb0eacda53741ea472a204acf26054/pkg/events/redis/store.go#L378-L388